File: NYTPhotoViewer/NYTInterstitialViewController.m
//
//  NYTInterstitialViewController.m
//  NYTPhotoViewer
//
//  Created by Howarth, Craig on 4/17/18.
//  Copyright © 2018 NYTimes. All rights reserved.
//

#import "NYTInterstitialViewController.h"
#import "NYTPhoto.h"

@interface NYTInterstitialViewController ()

@property (nonatomic, nullable) id <NYTPhoto> photo;
@property (nonatomic, nullable) UIView *interstitialView;
@property (nonatomic) NSUInteger photoViewItemIndex;
@property (nonatomic, nullable) NSArray<NSLayoutConstraint *> *constraints;

- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_DESIGNATED_INITIALIZER;

@end

@implementation NYTInterstitialViewController

#pragma mark - UIViewController

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    return [self initWithView:nil itemIndex:0];
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];

    if (self) {
        [self commonInitWithView:nil itemIndex:0];
    }

    return self;
}

- (void)prepareLayout {
    if (self.interstitialView.translatesAutoresizingMaskIntoConstraints) {
        self.interstitialView.center = CGPointMake(CGRectGetMidX(self.view.bounds), CGRectGetMidY(self.view.bounds));
    } else if (self.constraints == NULL) {
        self.constraints = @[
          [self.interstitialView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
          [self.interstitialView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
        ];
        [NSLayoutConstraint activateConstraints:self.constraints];
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self.view addSubview:self.interstitialView];
    [self prepareLayout];
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    [self prepareLayout];
}

- (BOOL)prefersHomeIndicatorAutoHidden {
    return YES;
}

#pragma mark - NYTPhotoViewController

- (instancetype)initWithView:(UIView *)interstitialView itemIndex:(NSUInteger)itemIndex {
    self = [super initWithNibName:nil bundle:nil];

    if (self) {
        [self commonInitWithView:interstitialView itemIndex:itemIndex];
    }

    return self;
}

- (void)commonInitWithView:(UIView *)interstitialView itemIndex:(NSUInteger)itemIndex {
    _photo = nil;
    _photoViewItemIndex = itemIndex;
    _interstitialView = interstitialView;
}

@end

File: Sources/NYTPhotoViewer/NYTInterstitialViewController.swift
//
//  NYTInterstitialViewController.swift
//  NYTPhotoViewer
//
//  Created by Howarth, Craig on 4/17/18.
//  Copyright © 2018 NYTimes. All rights reserved.
//

import UIKit

/// The view controller controlling the display of an interstitial view.
@MainActor
final class NYTInterstitialViewController: UIViewController, NYTPhotoViewerContainer {
    
    // MARK: - NYTPhotoViewerContainer
    
    let photo: (any NYTPhoto)? = nil
    let interstitialView: UIView?
    let photoViewItemIndex: Int
    
    private var constraints: [NSLayoutConstraint]?
    
    // MARK: - Initialization
    
    /// The designated initializer that takes an interstitial view.
    ///
    /// - Parameters:
    ///   - view: The view object that this view controller manages.
    ///   - itemIndex: The index of this view controller in the photo viewer collection.
    init(view: UIView?, itemIndex: Int) {
        self.interstitialView = view
        self.photoViewItemIndex = itemIndex
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        self.interstitialView = nil
        self.photoViewItemIndex = 0
        super.init(coder: coder)
    }
    
    // MARK: - View Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let interstitialView = interstitialView {
            view.addSubview(interstitialView)
        }
        prepareLayout()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        prepareLayout()
    }
    
    override var prefersHomeIndicatorAutoHidden: Bool {
        true
    }
    
    // MARK: - Private Methods
    
    private func prepareLayout() {
        guard let interstitialView = interstitialView else { return }
        
        if interstitialView.translatesAutoresizingMaskIntoConstraints {
            interstitialView.center = CGPoint(x: view.bounds.midX, y: view.bounds.midY)
        } else if constraints == nil {
            let newConstraints = [
                interstitialView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                interstitialView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
            ]
            NSLayoutConstraint.activate(newConstraints)
            self.constraints = newConstraints
        }
    }
}

